### DISCUSSION TITLE
Ways to speed up Wagtail installs

### DISCUSSION THREAD

**Discussion by thibaudcolas** [wagtail/wagtail#12724] [Development]
### Is your proposal related to a problem?

I just came out of a workshop where we had participant install Wagtail on Windows, and we noticed quite slow install times, up to 5-10 minutes for a participant. I suppose an element of this might just be slower machines, and @tomdyson let me know on Slack that it’s also important to use an up-to-date version of pip as older versions might not be able to use wheels. I wonder if there are other things we could do to speed up the install though, even if only by a few seconds.

### Describe the solution you'd like

Inspecting a Wagtail install (on macOS),

#### Wheels

It looks like three of Wagtail’s dependencies don’t have prepublished wheels. Should we consider making PRs?

```
Building wheel for l18n (setup.py) ... done
Building wheel for django-treebeard (setup.py) ... done
Building wheel for draftjs-exporter (setup.py) ... done
```

For `draftjs-exporter` – I can easily publish as a wheel, not sure why it’s not set up yet (https://github.com/springload/draftjs_exporter/issues/132).

#### static_src

There is a line in our MANIFEST.in excluding a static_src folder – however it’s using a path that no longer exists since wagtail 2.0. Should we have all `static_src` folders there?

https://github.com/wagtail/wagtail/blob/e10a9f26e7c944fd3a037a0459d371ab122ff96d/MANIFEST.in#L3

```
global-exclude static_src
```

#### tests

We could also probably not publish our tests with Wagtail, hopefully just with:

```
global-exclude tests
global-exclude tests.py
```

#### Source maps

Wagtail only ships three source maps, but together those files account for about 4MB. I think I liked the idea of shipping these with Wagtail so users could report on JS errors that were intelligible, but in practice I haven’t seen this done that much since we started shipping those files.

```
2.1M	admin/static/wagtailadmin/js/draftail.js.map
1.1M	admin/static/wagtailadmin/js/vendor.js.map
440K	admin/static/wagtailadmin/js/wagtailadmin.js.map
```

To exclude these,

```
global-exclude *.map
```

#### Other files

Here are other patterns we should be able to exclude, resulting in small savings:

```
global-exclude .gitignore
# Single file but is quite big and does nothing at runtime.
global exclude wagtail-icomoon.json
```

(Describe your proposed solution here.)

---

Here is the change with all those modifications done:

```
# Before
29MB, 1742 directories, 2803 files
# After
19MB, 1647 directories, 2356 files
```

Not night-and-day, but worth considering. Is there anything I’ve missed? Anything from the list above that seems like a bad idea?

**Comment by allcaps** [comment]
```
# List files ordered by size.
pip install wagtail
cd site-packages/wagtail
find . -type f -print0 | xargs -0 ls -l | sort -k5,5rn
```

- `./contrib/table_block/static/table_block/js/vendor/handsontable-6.2.2.full.min.js` 0.8MB seems big. Isn't there _also_ a license problem?
- `./node_modules/` ?
- `*.woff`, and `*.woff2` 3.9 MB. Font families (Roboto and OpenSans), various weights (light, regular, semi-bold, bold) and various styles (roman, italic, slab). Are all fonts used in the ui?

**Comment by gasman** [comment]
There's no licence problem with handsontable - that was a miscommunication. The non-OSS-compatible licence was introduced in version 7, so as long as we stick with 6.x I believe we're fine.

**Comment by thibaudcolas** [comment]
Thanks @allcaps, those commands help. We only need `.woff` files for IE11 so I imagine that will be a nice saving once we can stop supporting it. For other font files, it might be worth reviewing which exactly are used, although install size really isn’t the main concern here. `node_modules` shouldn’t be published, it doesn’t seem to be from what I can see.

**Comment by Stormheg** [comment]
I've opened a PR to add wheel distributions to draftjs_exporter: https://github.com/springload/draftjs_exporter/pull/133

**Comment by thibaudcolas** [comment]
draftjs_exporter is now published with wheels (thank you Storm!), and I’ve opened an issue for https://github.com/django-treebeard/django-treebeard/issues/169. I’m not entirely sure whether https://github.com/tkhyn/l18n can also be changed to be published as a wheel so have held off for now.

For this issue, judging by the emoji responses it looks like what I proposed above seems like a good idea – so I’d welcome a PR to implement the suggested `MANIFEST.in` patterns:

```
global-exclude static_src
global-exclude tests
global-exclude tests.py
global-exclude *.map
global-exclude .gitignore
global exclude wagtail-icomoon.json
```

…with relevant comments explaining what each pattern does.

This is quite a simple change … so tagging as "good first issue" if anyone wants to pick it up, although you do need to know a fair bit about Python packaging to properly test those changes :-)

**Comment by merwok** [comment]
Please don’t exclude tests from sdists!  Redistributors (packagers for free operating sytems like Debian or Fedora) use sdists for their sources and want to run tests as part of the packaging process.  (Even if wagtail is not packaged for now, it could be, and also popular projects are often copied as examples by other developers.)

Tests don’t have to be in wheels, and don’t have to be installed, both of which are achieved by not listing the tests as `packages` or `py_modules`.  Some projects do that to let end-users run the test suite when they have other issues, but in that case the tests should be a sub-package of the project. I see wagtail does use sub-packages, and because you’re using `setuptools.find_packages()` they are going to be installed, unless you exclude them https://setuptools.readthedocs.io/en/latest/setuptools.html#using-find-packages — I am not sure if using `exclude` to avoid installing will also prevent the sub-packages from being included in sdists, but if that’s the case, add some `include` directives would correct that.

**Comment by thibaudcolas** [comment]
Hey @merwok, thank you for providing input on this. I think I’m very happy for tests to be there if anyone does use them – but to be honest I’m not convinced anyone does. If the tests can be absent from wheels only, I imagine it solves the problem though for the "install speed" factor, assuming we can expect people installing Wagtail to get wheels by default. I didn’t realise it was that simple for a wheel to have different content from the source distribution.

> Even if wagtail is not packaged for now, it could be, and also popular projects are often copied as examples by other developers.

This is just my personal opinion but I don’t really see the appeal in having Wagtail re-packaged elsewhere. We make releases very frequently, so for end users using other channels than PyPI means an increased likelihood of being stuck with outdated releases, and for us as maintainers it also means having to support people on older releases than we’d like. For example [anaconda’s Wagtail](https://github.com/conda-forge/wagtail-feedstock) is stuck on v2.7. [aldryn-wagtail](https://github.com/divio/aldryn-wagtail) doesn’t publish patch releases for our LTS versions.

> Some projects do that to let end-users run the test suite when they have other issues

Is the Wagtail test suite runnable currently? It’s missing its `runtests.py` in the sdist at least. I have yet to see a project that has this as a documented troubleshooting step (examples welcome), and people who would be knowledgeable enough to do this could just as well install from git I think.

**Comment by thibaudcolas** [comment]
Both draftjs_exporter (https://github.com/springload/draftjs_exporter/issues/132) and django-treebeard (https://github.com/django-treebeard/django-treebeard/issues/169) now use wheels. l18n doesn’t but @vsalvino requested it in https://github.com/tkhyn/l18n/issues/7.

Coming back to https://github.com/wagtail/wagtail/issues/6130#issue-634994794, I would welcome PRs to:

- Re-exclude `static_src` properly.
- Exclude source maps.
- Exclude wagtail-icomoon.json

In both cases we’ll want to make sure the files are correctly excluded in both source and wheel distributions.

For other options outlined above like the tests – I’d be happy to see someone try to remove them in a PR, but I think we still need further validation that this is indeed useful to do.

**Comment by thibaudcolas** [comment]
I went ahead to implement what I suggested above in #8030, alongside a related packaging fix. To me this is enough of an improvement that we can close this issue afterwards, unless there is strong consensus it’s worthwhile for us to exclude test code (`tests/` folders and `tests.py` files) from published artifacts.

@allcaps @Stormheg @gasman would one of you be ok to review #8030?

**Comment by zerolab** [comment]
Out of curiosity (and only if you have time), how do the packaged versions look like with all tests excluded?

**Comment by thibaudcolas** [comment]
I tried the following patterns, which made the source dist about 20% smaller (from 5.5MB to 4.4):

```
prune **/tests
global-exclude tests.py
```

Some / all of the test code was still present in the wheel. I assumed this was because of our `find_packages` usage, but didn’t investigate.

**Comment by zerolab** [comment]
imho that is a change worth doing. What do others think?

**Comment by gasman** [comment]
> imho that is a change worth doing. What do others think?

Seems good to me, but let's make sure we're not also dropping modules that end-user code is relying on for testing, e.g. [wagtail.tests.utils](https://docs.wagtail.org/en/stable/advanced_topics/testing.html).

**Comment by zerolab** [comment]
> Some / all of the test code was still present in the wheel. I assumed this was because of our find_packages usage, but didn’t investigate.

it is indeed. wheels doesn't play nice with setup tools/ find_packages excludes

**Comment by thibaudcolas** [comment]
👍  marking this as a `good first issue` again, for someone to update our `MANIFEST.in` and `setup.py` as described above so our package artifacts don’t contain test code.

If no one picks this up I’ll get to it eventually, but not as part of #8030.

**Comment by vsalvino** [comment]
> Seems good to me, but let's make sure we're not also dropping modules that end-user code is relying on for testing, e.g. [wagtail.tests.utils](https://docs.wagtail.org/en/stable/advanced_topics/testing.html).

Yes, this is important. However the actual tests which test wagtail itself could be excluded. In our projects we usually move those outside of the package folder, and have it import the package, to further separate the code.

**Comment by PaarthAgarwal** [comment]
I'd like to work on this issue

**Comment by kaedroho** [comment]
> I just came out of a workshop where we had participant install Wagtail on Windows, and we noticed quite slow install times, up to 5-10 minutes for a participant.

Do we know if this was caused by a large package size? downloading an extracting a 29MB file shouldn't take anywhere near that long on a modern machine with a reasonable internet connection, so I wonder if something else was going on.

Maybe it had to compile something like psycopg2 or Pillow?

Was migrations included in this time? We can definitely optimise those a lot.

**Comment by thibaudcolas** [comment]
The "5-10" minutes quoted was just `pip install wagtail`, I don’t recall pip saying exactly which parts of the install took time.

In beginner-level workshops for people who aren’t in tech, it’s very common for participants to have work laptops with very low specs (Windows budget laptop from 10 years ago with a HDD that might have never had a defrag). It’s also fairly common for those people to have patchy internet connections. 

If there are other ways to speed up the beginner experience I think they’d be well worth looking into, the initial list up there is just what seemed like obvious improvements at the time.

**Comment by lb-** [comment]
Flagging https://github.com/wagtail/wagtail/pull/8812 -> some consolidation of Excel-ish libraries should speed things up somewhat.

**Comment by ArijeetBanerjee** [comment]
I want to work on this issue. kindly give me permission and access to work on this issue

**Comment by thibaudcolas** [comment]
@ArijeetBanerjee It’s very rude to claim you want to work on everything all at the same time, "good first issues" are there for all newcomers to pick and choose. If you want to work on an issue, pick one that’s not already been claimed by someone else, and show us what you get on with.

**Comment by ArijeetBanerjee** [comment]
@thibaudcolas  sorry my bad

**Comment by thibaudcolas** [comment]
I have a feeling this issue has outlived its usefulness a bit. We still want to speed up Wagtail installs of course, but lots of the changes initially discussed here are already done – and I’m sure there are a lot of other possible changes that we could identify.

I’ll convert this to a discussion now, if anyone feels like there’s more to do please share your thoughts!

**Comment by vsalvino** [comment]
I do think there is one remaining performance improvement needed for new installations, and that would be squashing migrations. There are so many migrations that they take quite a while to run on the first install. Of course - this is entirely subjective, and not scientific at all - but hopefully this provides a little insight.

Using Powershell on Windows 11 (should be sufficiently slow!) I ran some of the following measurements using powershell's `Measure-Command`
| command | time |
| --- | --- |
| `pip install wagtail --no-cache-dir ` | 15 seconds |
| `wagtail startproject slowtail .` | 1.5 seconds |
| SQLite: `python manage.py migrate` | 7 seconds |
| Local MySQL:  `python manage.py migrate` | 22 seconds |
| Remote MySQL: `python manage.py migrate` | 70 seconds |

On CodeRed Cloud, we also measure the total time it takes to initialize a wagtail site as well, on every single deployment.

| CodeRed Cloud configuration | time |
| --- | --- |
| Python 3.13.1, Wagtail 6.3.1, MySQL | 63 seconds |
| Python 3.13.1, Wagtail 6.3.1, SQLite | 36 seconds |

We do provide a local cache of all wagtail dependencies on the file system, which empirically sped up the pip install process by an order of magnitude. So the real performance issue here is definitely the migrations.

Below is the output log of a new Wagtail install on CodeRed Cloud using MySQL (our most popular database), including the time measurement at the end.

```
> Initializing on: Fri Dec 20 17:26:17 UTC 2024
> Defaulting to user installation because normal site-packages is not writeable
> Looking in indexes: https://pypi.org/simple, file:///var/cache/pypi/
> Processing /var/cache/pypi/draftjs-exporter/draftjs_exporter-5.0.0-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/djangorestframework/djangorestframework-3.15.2-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/django-filter/django_filter-24.3-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/django/Django-5.1.4-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/wagtail/wagtail-6.3.1-py3-none-any.whl
> Processing /var/cache/pypi/django-permissionedforms/django_permissionedforms-0.1-py2.py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/django-taggit/django_taggit-6.1.0-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/django-modelcluster/django_modelcluster-6.4-py2.py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/django-treebeard/django_treebeard-4.7.1-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/six/six-1.17.0-py2.py3-none-any.whl (from l18n>=2018.5->wagtail==6.3.*)
> Processing /var/cache/pypi/idna/idna-3.10-py3-none-any.whl (from requests<3.0,>=2.11.1->wagtail==6.3.*)
> Processing /var/cache/pypi/beautifulsoup4/beautifulsoup4-4.12.3-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/requests/requests-2.32.3-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/sqlparse/sqlparse-0.5.3-py3-none-any.whl (from Django<6.0,>=4.2->wagtail==6.3.*)
> Processing /var/cache/pypi/et-xmlfile/et_xmlfile-2.0.0-py3-none-any.whl (from openpyxl<4.0,>=3.0.10->wagtail==6.3.*)
> Processing /var/cache/pypi/laces/laces-0.1.1-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/pytz/pytz-2024.2-py2.py3-none-any.whl (from l18n>=2018.5->wagtail==6.3.*)
> Processing /var/cache/pypi/soupsieve/soupsieve-2.6-py3-none-any.whl (from beautifulsoup4<4.13,>=4.8->wagtail==6.3.*)
> Processing /var/cache/pypi/willow/willow-1.9.0-py3-none-any.whl (from Willow[heif]<2,>=1.8.0->wagtail==6.3.*)
> Processing /var/cache/pypi/anyascii/anyascii-0.3.2-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/charset-normalizer/charset_normalizer-3.4.0-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (from requests<3.0,>=2.11.1->wagtail==6.3.*)
> Processing /var/cache/pypi/l18n/l18n-2021.3-py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/openpyxl/openpyxl-3.1.5-py2.py3-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/telepath/telepath-0.3.1-py38-none-any.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/pillow/pillow-11.0.0-cp313-cp313-manylinux_2_28_x86_64.whl (from wagtail==6.3.*)
> Processing /var/cache/pypi/asgiref/asgiref-3.8.1-py3-none-any.whl (from Django<6.0,>=4.2->wagtail==6.3.*)
> Processing /var/cache/pypi/defusedxml/defusedxml-0.7.1-py2.py3-none-any.whl (from Willow<2,>=1.8.0->Willow[heif]<2,>=1.8.0->wagtail==6.3.*)
> Processing /var/cache/pypi/urllib3/urllib3-2.2.3-py3-none-any.whl (from requests<3.0,>=2.11.1->wagtail==6.3.*)
> Installing collected packages: pytz, filetype, draftjs-exporter, urllib3, telepath, sqlparse, soupsieve, six, Pillow, idna, et-xmlfile, defusedxml, charset-normalizer, certifi, asgiref, anyascii, Willow, requests, pillow-heif, openpyxl, l18n, Django, beautifulsoup4, laces, djangorestframework, django-treebeard, django-taggit, django-permissionedforms, django-modelcluster, django-filter, wagtail
> Downloading pillow_heif-0.21.0-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.8 MB)
> WARNING: Location 'file:///var/cache/pypi/pillow-heif/' is ignored: it is neither a file nor a directory.
>   Downloading pillow_heif-0.21.0-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (9.8 kB)
> Processing /var/cache/pypi/certifi/certifi-2024.12.14-py3-none-any.whl (from requests<3.0,>=2.11.1->wagtail==6.3.*)
> Processing /var/cache/pypi/filetype/filetype-1.2.0-py2.py3-none-any.whl (from Willow<2,>=1.8.0->Willow[heif]<2,>=1.8.0->wagtail==6.3.*)
> Collecting pillow-heif<1.0.0,>=0.13.0 (from Willow[heif]<2,>=1.8.0->wagtail==6.3.*)
> Successfully installed Django-5.1.4 Pillow-11.0.0 Willow-1.9.0 anyascii-0.3.2 asgiref-3.8.1 beautifulsoup4-4.12.3 certifi-2024.12.14 charset-normalizer-3.4.0 defusedxml-0.7.1 django-filter-24.3 django-modelcluster-6.4 django-permissionedforms-0.1 django-taggit-6.1.0 django-treebeard-4.7.1 djangorestframework-3.15.2 draftjs-exporter-5.0.0 et-xmlfile-2.0.0 filetype-1.2.0 idna-3.10 l18n-2021.3 laces-0.1.1 openpyxl-3.1.5 pillow-heif-0.21.0 pytz-2024.2 requests-2.32.3 six-1.17.0 soupsieve-2.6 sqlparse-0.5.3 telepath-0.3.1 urllib3-2.2.3 wagtail-6.3.1
> Creating a Wagtail project called slowtail using the default Wagtail template
> Success! slowtail has been created
> System check identified some issues:
> WARNINGS:
> wagtailcore.WorkflowState: (models.W036) MySQL does not support unique constraints with conditions.
> Running migrations:
> Operations to perform:
> 	HINT: A constraint won't be created. Silence this warning if you don't care about it.
>
>   Apply all migrations: admin, auth, contenttypes, home, sessions, taggit, wagtailadmin, wagtailcore, wagtaildocs, wagtailembeds, wagtailforms, wagtailimages, wagtailredirects, wagtailsearch, wagtailusers
>   Applying contenttypes.0001_initial... OK
>   Applying admin.0002_logentry_remove_auto_add... OK
>   Applying admin.0003_logentry_add_action_flag_choices... OK
>   Applying contenttypes.0002_remove_content_type_name... OK
>   Applying admin.0001_initial... OK
>   Applying auth.0001_initial... OK
>   Applying auth.0005_alter_user_last_login_null... OK
>   Applying auth.0006_require_contenttypes_0002... OK
>   Applying auth.0007_alter_validators_add_error_messages... OK
>   Applying auth.0010_alter_group_name_max_length... OK
>   Applying auth.0004_alter_user_username_opts... OK
>   Applying auth.0008_alter_user_username_max_length... OK
>   Applying auth.0003_alter_user_email_max_length... OK
>   Applying auth.0012_alter_user_first_name_max_length... OK
>   Applying auth.0009_alter_user_last_name_max_length... OK
>   Applying auth.0002_alter_permission_name_max_length... OK
>   Applying auth.0011_update_proxy_permissions... OK
>   Applying wagtailcore.0019_verbose_names_cleanup... OK
>   Applying wagtailcore.0023_alter_page_revision_on_delete_behaviour... OK
>   Applying wagtailcore.0022_add_site_name... OK
>   Applying wagtailcore.0021_capitalizeverbose... OK
>   Applying wagtailcore.0025_collection_initial_data... OK
>   Applying wagtailcore.0024_collection... OK
>   Applying wagtailcore.0001_squashed_0016_change_page_url_path_to_text_field... OK
>   Applying wagtailcore.0017_change_edit_page_permission_description... OK
>   Applying wagtailcore.0018_pagerevision_submitted_for_moderation_index... OK
>   Applying wagtailcore.0020_add_index_on_page_first_published_at... OK
>   Applying wagtailcore.0028_merge... OK
>   Applying wagtailcore.0027_fix_collection_path_collation... OK
>   Applying wagtailcore.0026_group_collection_permission... OK
>   Applying wagtailcore.0024_alter_page_content_type_on_delete_behaviour... OK
>   Applying wagtailcore.0029_unicode_slugfield_dj19... OK
>   Applying wagtailcore.0030_index_on_pagerevision_created_at... OK
>   Applying wagtailcore.0036_populate_page_last_published_at... OK
>   Applying wagtailcore.0038_make_first_published_at_editable... OK
>   Applying wagtailcore.0035_page_last_published_at... OK
>   Applying wagtailcore.0033_remove_golive_expiry_help_text... OK
>   Applying wagtailcore.0034_page_live_revision... OK
>   Applying wagtailcore.0031_add_page_view_restriction_types... OK
>   Applying wagtailcore.0037_set_page_owner_editable... OK
>   Applying wagtailcore.0032_add_bulk_delete_page_permission... OK
>   Applying wagtailcore.0039_collectionviewrestriction... OK
>   Applying home.0002_create_homepage... OK
>   Applying taggit.0002_auto_20150616_2121... OK
>   Applying home.0001_initial... OK
>   Applying sessions.0001_initial... OK
>   Applying taggit.0001_initial... OK
>   Applying wagtailcore.0040_page_draft_title... OK
>   Applying taggit.0003_taggeditem_add_unique_index... OK
>   Applying wagtailcore.0041_group_collection_permissions_verbose_name_plural... OK
>   Applying wagtailadmin.0001_create_admin_access_permissions... OK
>   Applying wagtailcore.0043_lock_fields... OK
>   Applying wagtailadmin.0002_admin... OK
>   Applying taggit.0006_rename_taggeditem_content_type_object_id_taggit_tagg_content_8fc721_idx... OK
>   Applying wagtailadmin.0003_admin_managed... OK
>   Applying wagtailadmin.0004_editingsession... OK
>   Applying taggit.0004_alter_taggeditem_content_type_alter_taggeditem_tag... OK
>   Applying wagtailadmin.0005_editingsession_is_editing... OK
>   Applying wagtailcore.0042_index_on_pagerevision_approved_go_live_at... OK
>   Applying taggit.0005_auto_20220424_2025... OK
>   Applying wagtailcore.0044_add_unlock_grouppagepermission... OK
>   Applying wagtailcore.0045_assign_unlock_grouppagepermission... OK
>   Applying wagtailcore.0046_site_name_remove_null... OK
>   Applying wagtailcore.0049_taskstate_finished_by... OK
>   Applying wagtailcore.0051_taskstate_comment... OK
>   Applying wagtailcore.0047_add_workflow_models... OK
>   Applying wagtailcore.0048_add_default_workflows... OK
>   Applying wagtailcore.0050_workflow_rejected_to_needs_changes... OK
>   Applying wagtailcore.0052_pagelogentry... OK
>   Applying wagtailcore.0053_locale_model... OK
>   Applying wagtailcore.0054_initial_locale... OK
>   Applying wagtailcore.0056_page_locale_fields_populate... OK
>   Applying wagtailcore.0055_page_locale_fields... OK
>   Applying wagtailcore.0057_page_locale_fields_notnull... OK
>   Applying wagtailcore.0061_change_promote_tab_helpt_text_and_verbose_names... OK
>   Applying wagtailcore.0060_fix_workflow_unique_constraint... OK
>   Applying wagtailcore.0059_apply_collection_ordering... OK
>   Applying wagtailcore.0058_page_alias_of... OK
>   Applying wagtailcore.0062_comment_models_and_pagesubscription... OK
>   Applying wagtailcore.0064_log_timestamp_indexes... OK
>   Applying wagtailcore.0063_modellogentry... OK
>   Applying wagtailcore.0068_log_entry_empty_object... OK
>   Applying wagtailcore.0066_collection_management_permissions... OK
>   Applying wagtailcore.0069_log_entry_jsonfield... OK
>   Applying wagtailcore.0070_rename_pagerevision_revision... OK
>   Applying wagtailcore.0071_populate_revision_content_type... OK
>   Applying wagtailcore.0072_alter_revision_content_type_notnull... OK
>   Applying wagtailcore.0074_revision_object_str... OK
>   Applying wagtailcore.0077_alter_revision_user... OK
>   Applying wagtailcore.0073_page_latest_revision... OK
>   Applying wagtailcore.0078_referenceindex... OK
>   Applying wagtailcore.0075_populate_latest_revision_and_revision_object_str... OK
>   Applying wagtailcore.0076_modellogentry_revision... OK
>   Applying wagtailcore.0079_rename_taskstate_page_revision... OK
>   Applying wagtailcore.0080_generic_workflowstate... OK
>   Applying wagtailcore.0081_populate_workflowstate_content_type... OK
>   Applying wagtailcore.0082_alter_workflowstate_content_type_notnull... OK
>   Applying wagtailcore.0086_populate_grouppagepermission_permission... OK
>   Applying wagtailcore.0084_add_default_page_permissions... OK
>   Applying wagtailcore.0085_add_grouppagepermission_permission... OK
>   Applying wagtailcore.0083_workflowcontenttype... OK
>   Applying wagtailcore.0088_fix_log_entry_json_timestamps... OK
>   Applying wagtailcore.0089_log_entry_data_json_null_to_object... OK
>   Applying wagtailcore.0087_alter_grouppagepermission_unique_together_and_more... OK
>   Applying wagtailcore.0094_alter_page_locale... OK
>   Applying wagtailcore.0091_remove_revision_submitted_for_moderation... OK
>   Applying wagtailcore.0092_alter_collectionviewrestriction_password_and_more... OK
>   Applying wagtailcore.0093_uploadedfile... OK
>   Applying wagtailcore.0090_remove_grouppagepermission_permission_type... OK
>   Applying wagtaildocs.0002_initial_data... OK
>   Applying wagtaildocs.0003_add_verbose_names... OK
>   Applying wagtaildocs.0001_initial... OK
>   Applying wagtaildocs.0005_document_collection... OK
>   Applying wagtaildocs.0004_capitalizeverbose... OK
>   Applying wagtaildocs.0006_copy_document_permissions_to_collections... OK
>   Applying wagtaildocs.0007_merge... OK
>   Applying wagtaildocs.0011_add_choose_permissions... OK
>   Applying wagtaildocs.0014_alter_document_file_size... OK
>   Applying wagtaildocs.0013_delete_uploadeddocument... OK
>   Applying wagtaildocs.0008_document_file_size... OK
>   Applying wagtaildocs.0009_document_verbose_name_plural... OK
>   Applying wagtailembeds.0002_add_verbose_names... OK
>   Applying wagtailembeds.0001_initial... OK
>   Applying wagtailembeds.0003_capitalizeverbose... OK
>   Applying wagtailembeds.0006_add_embed_hash... OK
>   Applying wagtaildocs.0012_uploadeddocument... OK
>   Applying wagtailembeds.0004_embed_verbose_name_plural... OK
>   Applying wagtailembeds.0005_specify_thumbnail_url_max_length... OK
>   Applying wagtaildocs.0010_document_file_hash... OK
>   Applying wagtailembeds.0007_populate_hash... OK
>   Applying wagtailforms.0002_add_verbose_names... OK
>   Applying wagtailembeds.0008_allow_long_urls... OK
>   Applying wagtailforms.0004_add_verbose_name_plural... OK
>   Applying wagtailembeds.0009_embed_cache_until... OK
>   Applying wagtailforms.0003_capitalizeverbose... OK
>   Applying wagtailforms.0001_initial... OK
>   Applying wagtailforms.0005_alter_formsubmission_form_data... OK
>   Applying wagtailimages.0027_image_description... OK
>   Applying wagtailredirects.0001_initial... OK
>   Applying wagtailimages.0001_squashed_0021... OK
>   Applying wagtailimages.0023_add_choose_permissions... OK
>   Applying wagtailimages.0022_uploadedimage... OK
>   Applying wagtailimages.0025_alter_image_file_alter_rendition_file... OK
>   Applying wagtailimages.0026_delete_uploadedimage... OK
>   Applying wagtailimages.0024_index_image_file_hash... OK
>   Applying wagtailredirects.0007_add_autocreate_fields... OK
>   Applying wagtailredirects.0002_add_verbose_names... OK
>   Applying wagtailredirects.0006_redirect_increase_max_length... OK
>   Applying wagtailredirects.0005_capitalizeverbose... OK
>   Applying wagtailredirects.0003_make_site_field_editable... OK
>   Applying wagtailredirects.0004_set_unique_on_path_and_site... OK
>   Applying wagtailredirects.0008_add_verbose_name_plural... OK
>   Applying wagtailsearch.0001_initial... OK
>   Applying wagtailsearch.0003_remove_editors_pick... OK
>   Applying wagtailsearch.0004_querydailyhits_verbose_name_plural... OK
>   Applying wagtailsearch.0002_add_verbose_names... OK
>   Applying wagtailsearch.0005_create_indexentry... OK
>   Applying wagtailsearch.0007_delete_editorspick... OK
>   Applying wagtailsearch.0006_customise_indexentry... OK
>   Applying wagtailusers.0003_add_verbose_names... OK
>   Applying wagtailsearch.0008_remove_query_and_querydailyhits_models... OK
>   Applying wagtailusers.0002_add_verbose_name_on_userprofile... OK
>   Applying wagtailusers.0001_initial... OK
>   Applying wagtailusers.0004_capitalizeverbose... OK
>   Applying wagtailusers.0005_make_related_name_wagtail_specific... OK
>   Applying wagtailusers.0014_userprofile_contrast... OK
>   Applying wagtailusers.0010_userprofile_updated_comments_notifications... OK
>   Applying wagtailusers.0009_userprofile_verbose_name_plural... OK
>   Applying wagtailusers.0007_userprofile_current_time_zone... OK
>   Applying wagtailusers.0006_userprofile_prefered_language... OK
>   Applying wagtailusers.0008_userprofile_avatar... OK
>   Applying wagtailusers.0011_userprofile_dismissibles... OK
>   Applying wagtailusers.0012_userprofile_theme... OK
>   Applying wagtailusers.0013_userprofile_density... OK
> 
> 221 static files copied to '/var/www/static', 221 post-processed.
> Python 3.13.1
> Initialized in 63 seconds.
```

**Reply by thibaudcolas** [reply]
Excellent! Is that something we can propose as a more specific issue then? Is it pretty clear how / how often to go about this or does it require a lot of discussion / prep?

**Reply by gasman** [reply]
For the record: https://github.com/wagtail/wagtail/issues/4965#issuecomment-503570537

...although I have to admit that my stance has softened since then. We just need to be aware that once we've squashed migrations once, we won't be able to do it again until enough time has passed that we're prepared to delete the previous round of unsquashed migrations (and thus break the ability to upgrade directly from a version that pre-dates the squashed migration to a version that post-dates the removal of unsquashed migrations).

**Reply by gasman** [reply]
Also see https://github.com/wagtail/wagtail/issues/5067#issuecomment-503584658 for the kind of issues that squashed migrations can cause when a project's migrations have dependencies somewhere in the middle of the squashed range.

**Reply by JoGorska** [reply]
I've never added a package to a project that would require so many migrations. There are 175 migrations, as per the list posted above. Not only does it take a long time to apply, but it also takes a long time to get deployed. I am testing migrations in the pipeline, and as you can imagine, this took far longer than normal to go through.

The number of migrations is one thing, but also, among those migrations, there are a few that are non-reversible migrations. I raised [ticket](https://github.com/wagtail/wagtail/issues/12834 ) to describe the problem. My pipeline failed on that. 

You really have to weigh the pros and cons of squashing migrations, but it would solve the problem of installation for some time. It will never eliminate it to 1 or 2, because there are lots of apps inside the project; having said that, it would be a meaningful improvement. 

I'm happy to write a ticket and contribute if it helps. It needs to be planned well, taking into consideration a good deployment plan and widespread information.

**Comment by lb-** [comment]
See https://github.com/wagtail/wagtail/pull/8112 for an initial attempt to remove tests and published artifacts from the install process.

This had a good start with some useful feedback in reviews but the PR went stale. It would be great if anyone is keen to pick up where it left off.

I have not raised a new issue for this as it seems we are just covering these potential sub-issues / tasks in this discussion thread for now.
